### PR TITLE
[AMD][Bugfix][EPLB] Fix elastic EP scaling accuracy on ROCm

### PR DESCRIPTION
**Pull request by okorzh-amd**
## Purpose

Two fixes for distributed/test_elastic_ep.py, which collapsed GSM8K accuracy at the initial and scaled checkpoints on ROCm:

Bug 1 Scale-up produces garbage output on ROCm (accuracy=0)

Root cause. New elastic-EP ranks are created with load_dummy_weights=True. On ROCm, initialize_single_dummy_weight() .zero_()s every non-floating-point tensor in model.state_dict(). The FusedMoE expert-topology buffers _expert_map, expert_mask, and expert_{global_to_physical,physical_to_global,local_to_global} were registered as persistent integer buffers, so they were zeroed right after being built correctly for the new EP topology. That corrupted all2all dispatch garbage output. CUDA never zeroes integer tensors, so the bug was ROCm-only.

This corrects an earlier hypothesis: the map is not built unsharded at ep_size==1; it is built correctly at the target ep_size and then corrupted. (Thanks @itayalroy for identifying the dummy-init zeroing path.)

Fix. Register those five buffers with persistent=False. They are per-rank, deterministically-derived EP topology metadata not checkpoint state so they should never be in state_dict() (matches moondream3.py's _expert_map registration). This keeps them out of state_dict(), so:
- ROCm dummy-weight init can no longer zero them, and
- batch_transfer_weights() (which iterates state_dict()) can no longer clobber them from a sender rank.

As a result, no post-reshuffle metadata rebuild is needed and no name-based transfer filter for these tensors is needed.

Bug 2 Sync-EPLB accuracy loss from redundant rearranges

rearrange_expert_weights_inplace() was firing every step_interval even when the proposed map barely improved balance (imbalance 1.0), causing ~16s lateral all-to-all churn that starved the engine and dropped accuracy. We now compute per-rank load imbalance for the current vs proposed map and skip the rearrange when improvement is <5%, decided collectively via all-reduce and guarded to non-scale-up rearranges.

## Test Plan
run failed CI tests

## Test Result
```
distributed/test_elastic_ep.py::test_elastic_ep_scaling[sync_eplb]          PASSED
distributed/test_elastic_ep.py::test_elastic_ep_scaling[async_eplb]         PASSED
distributed/test_elastic_ep.py::test_elastic_ep_scaling_uneven[sync_eplb]   PASSED
distributed/test_elastic_ep.py::test_elastic_ep_scaling_uneven[async_eplb]  PASSED
================= 4 passed, 18 warnings in 1194.63s (0:19:54) ==================
```

Per-checkpoint accuracies (all >= 0.58 threshold; previously 0.19 / 0.016):

image

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/distributed/eplb/eplb_state.py [hunk 1]
```diff
@@ -825,23 +825,65 @@ def rearrange(
...
+                    rearrange_expert_weights_inplace(
+                        eplb_model_state.physical_to_logical_map,
+                        new_physical_to_logical_map,
+                        eplb_model_state.model.expert_weights,
+                        eplb_model_state.expert_buffer,
+                        ep_group,
+                        eplb_model_state.communicator,
+                        is_profile,
+                        rank_mapping,
+                    )
```
Comment (tlrmchlsmth) [inline vllm/distributed/eplb/eplb_state.py]: Can this cause deadlocks if one rank decides to rearrange and another decides not to, due to slight numerical differences when calculating the imbalance?

**Comment by tlrmchlsmth** [comment]
>~16-21s of tiny per-expert P2P

Wow that is very significant! What library are you using for EPLB transfers?

File: vllm/distributed/elastic_ep/elastic_execute.py [hunk 2]
```diff
@@ -550,6 +550,16 @@ def _perform_eplb_reshuffle(
...
+
+        if rank_mapping is None:  # scale-up
+            model = self.worker.model_runner.get_model()
+            if hasattr(model, "update_physical_experts_metadata"):
+                ep_size = get_ep_group().world_size
+                num_physical_experts = eplb_model_state.physical_to_logical_map.shape[1]
+                model.update_physical_experts_metadata(
+                    num_physical_experts=num_physical_experts,
+                    num_local_physical_experts=num_physical_experts // ep_size,
+                )
```
Comment (tlrmchlsmth) [inline vllm/distributed/elastic_ep/elastic_execute.py]: Why is this needed on on RoCM? Or is it a general bug?

File: vllm/distributed/eplb/eplb_state.py (hunk 1 above)
Comment (okorzh-amd) [reply vllm/distributed/eplb/eplb_state.py]: In the current design the decision cannot actually diverge, because every input to the skip test is rank-identical:
* global_expert_load_window (_gl) is the all-reduced load, so it's the same tensor on every rank.
* physical_to_logical_map is a replicated global invariant, and new_physical_to_logical_map comes from the deterministic policy.rebalance_experts(...) fed that same all-reduced load so it's identical on every rank too.
* _rank_imbalance is a pure function of those integer maps + scalars. Same inputs + same op sequence on the same device type leads to bit-identical float result. FP "slight numerical differences" only appear when the inputs differ across ranks, which they don't here.

So on a homogeneous cluster there's no split-brain. But that safety is an implicit invariant, and it would silently break on a heterogeneous cluster (mixed GPU archs) or if someone later feeds a non-all-reduced load. Since rearrange_expert_weights_inplace is a collective, a divergent decision would deadlock. Let me try a solution with reduce to see how much perf it is going to cost.

**Comment by okorzh-amd** [comment]
> Wow that is very significant! What library are you using for EPLB transfers?

@tlrmchlsmth torch_nccl is silently promoted to pynccl (see eplb_communicator.py ~L739-744) which actually is RCCL ncclSend/ncclRecv for the test.

File: vllm/distributed/elastic_ep/elastic_execute.py (hunk 2 above)
Comment (okorzh-amd) [reply vllm/distributed/elastic_ep/elastic_execute.py]: yes, it is shared path. 
@pavanimajety who would be a good PIC from Nvidia to review it?

**Comment by tlrmchlsmth** [comment]
> > Wow that is very significant! What library are you using for EPLB transfers?
> 
> @tlrmchlsmth torch_nccl is silently promoted to pynccl (see eplb_communicator.py ~L739-744) which actually is RCCL ncclSend/ncclRecv for the test.

I see - I asked because we switched the default EPLB communication library to be NIXL with a fallback to gloo. We found that NCCL can sometimes hang (due to a NCCL bug that was subsequently fixed), and NCCL is not compatible with the async case due to a lack of forward progress guarantees in CUDA. Is this the case in RCCL/ROCm as well?

**Comment by okorzh-amd** [comment]
> > > Wow that is very significant! What library are you using for EPLB transfers?
> > 
> > 
> > @tlrmchlsmth torch_nccl is silently promoted to pynccl (see eplb_communicator.py ~L739-744) which actually is RCCL ncclSend/ncclRecv for the test.
> 
> I see - I asked because we switched the default EPLB communication library to be NIXL with a fallback to gloo. We found that NCCL can sometimes hang (due to a NCCL bug that was subsequently fixed), and NCCL is not compatible with the async case due to a lack of forward progress guarantees in CUDA. Is this the case in RCCL/ROCm as well?

it's a small-message P2P performance problem, not the NCCL hang bug. So the answer to your question is yes, but this particular code is platform-agnostic.

File: vllm/distributed/eplb/eplb_state.py (hunk 1 above)
Comment (okorzh-amd) [reply vllm/distributed/eplb/eplb_state.py]: perf looks good. I've update the code to have all_reduce for the decision

**Comment by SageMoore** [comment]
> each sync rearrange is ~16-21s

@okorzh-amd this is in incredible amount of transfer time. Can you elaborate on exactly what workload gives you those times? Are they exclusive to multi-node or are you seeing those transfer times on single node as well? I'd be surprised if EPLB gives any performance boost with transfer times like that.

File: vllm/distributed/elastic_ep/elastic_execute.py (hunk 2 above)
Comment (tlrmchlsmth) [reply vllm/distributed/elastic_ep/elastic_execute.py]: @itayalroy would be the right one

Comment (itayalroy) [reply vllm/distributed/elastic_ep/elastic_execute.py]: Why is this metadata update needed? both EEP reconfiguration and EPLB reshuffle never change which physical slots each rank owns, new ranks should already have the correct metadata.

I think the issue actually comes from `batch_transfer_weights()`, it overwrites the new rank’s AITER `expert_mask` with `expert_mask` from an existing rank, because it excludes `expert_map` by name but it does not exclude `expert_mask` (and this is why we don't see an issue with CUDA btw). Can we try to exclude the ROCM expert metadata there as well?

We should really avoid deciding what to transfer based on tensor names and explicitly define which model state should be copied to the new workers, this `batch_transfer_weights` filter is very hacky, but since this is existing behavior we can do it in a separate PR (unless you are up to fixing it)

Comment (okorzh-amd) [reply vllm/distributed/elastic_ep/elastic_execute.py]: Good catch on the expert_mask filter gap. I've added it to the exclusion (it's genuinely per-rank metadata that shouldn't be transferred). But I tested dropping the update_physical_experts_metadata rebuild and relying only on the exclusion, and scale-up still produces garbage (GSM8K 0.008). The root cause isn't the clobber: new scale-up ranks build their FusedMoE before joining the new EP group, so determine_expert_map() runs at ep_size==1 and returns an unsharded map (every global expert local) which matches the "all-local" symptom, not a sender's shard. So the rebuild is doing the real work of resharding for the new topology; the expert_mask exclusion is a correctness/hygiene fix on top. All 4 test_elastic_ep.py params pass on MI355X.

Comment (itayalroy) [reply vllm/distributed/elastic_ep/elastic_execute.py]: New ranks start with the target DP size ([code](https://github.com/vllm-project/vllm/blob/d82c7ee277116738e676cc22da779fe71d3059ca/vllm/v1/engine/utils.py#L858-L871)). FusedMoE reads that size from the DP group ([code](https://github.com/vllm-project/vllm/blob/d82c7ee277116738e676cc22da779fe71d3059ca/vllm/model_executor/layers/fused_moe/layer.py#L44-L64)), calculates flattened TP size as DP × PCP × TP ([code](https://github.com/vllm-project/vllm/blob/d82c7ee277116738e676cc22da779fe71d3059ca/vllm/model_executor/layers/fused_moe/config.py#L1099-L1107)), and uses that value as ep_size ([code](https://github.com/vllm-project/vllm/blob/d82c7ee277116738e676cc22da779fe71d3059ca/vllm/model_executor/layers/fused_moe/config.py#L1218-L1232)). This ep_size is passed to determine_expert_map() ([code](https://github.com/vllm-project/vllm/blob/d82c7ee277116738e676cc22da779fe71d3059ca/vllm/model_executor/layers/fused_moe/expert_map_manager.py#L440-L453)), so this value is never 1 for new ranks (unless ofc target ep size is actually 1). Moreover, if new ranks were initialized with ep_size == 1, we would also see the problem on CUDA because this path is shared. Finally I also printed ep_size from determine_expert_map() on new ranks and it equals the target ep size, so new ranks initializing FusedMoE with ep_size == 1 is definitely not the RC here.

I think that with ROCm, the expert metadata is getting corrupted somewhere along the way. We already found one issue in batch weight transfer. Another likely issue is during model loading: new ranks load model with dummy weights ([code](https://github.com/vllm-project/vllm/blob/d82c7ee277116738e676cc22da779fe71d3059ca/vllm/distributed/elastic_ep/elastic_execute.py#L179-L192)), and specifically on ROCm this seems to zero every non-floating-point tensor in state_dict() ([code](https://github.com/vllm-project/vllm/blob/d82c7ee277116738e676cc22da779fe71d3059ca/vllm/model_executor/model_loader/weight_utils.py#L1273-L1312)). Since _expert_map and expert_mask are registered integer buffers, they are probably being zeroed there even though they shouldn't be (they are metadata, not weights). The CUDA path does not zero integer tensors, which would explain why we do not see the same issue there.

There may be other issues as well, but I think we should identify the actual RC rather than rely on this update expert metadata call without fully understanding what went wrong. Otherwise, now or in the future, other tensors (unrelated to expert metadata) may get corrupted by the same path and not repaired.

Comment (okorzh-amd) [reply vllm/distributed/elastic_ep/elastic_execute.py]: @itayalroy thanks for your comments. I've updated the code and PR description. Please review when you have time.

**Comment by okorzh-amd** [comment]
> @okorzh-amd this is in incredible amount of transfer time. Can you elaborate on exactly what workload gives you those times? Are they exclusive to multi-node or are you seeing those transfer times on single node as well? I'd be surprised if EPLB gives any performance boost with transfer times like that.

@SageMoore good question. These are single-node numbers, so it's not a multi-node fabric issue. Exact workload is the CI test itself (tests/distributed/test_elastic_ep.py::test_elastic_ep_scaling[sync_eplb]): DeepSeek-V2-Lite-Chat on one MI355X node, 4 GPUs, TP=1, DP=2, --enable-eplb sync mode, step_interval=10, window_size=5, num_redundant_experts=0. 26 MoE layers * 64 experts = 1664 physical expert slots.

**Review by AndreasKaratzas — APPROVED** [review]
LGTM

@tlrmchlsmth PTAL

File: vllm/distributed/eplb/eplb_state.py [hunk 3]
```diff
@@ -825,23 +825,73 @@ def rearrange(
...
+                    local_skip = (
+                        _old_imb <= 1e-6 or (_old_imb - _new_imb) / _old_imb < 0.05
                     )
+                    _vote = torch.tensor(
+                        [1 if local_skip else 0],
+                        device=_gl.device,
+                        dtype=torch.int32,
+                    )
+                    all_reduce(_vote, group=ep_group)
+                    skip_rearrange = int(_vote.item()) == _ep
```
Comment (tlrmchlsmth) [inline vllm/distributed/eplb/eplb_state.py]:
```suggestion
                      logical_loads = global_expert_load_window.float()
                      ep_size = ep_group.size()

                      def rank_load_imbalance(mapping: torch.Tensor) -> float:
                          mapping = mapping.to(
                              device=logical_loads.device,
                              dtype=torch.long,
                          )
                          replica_counts = torch.zeros_like(logical_loads)
                          replica_counts.scatter_add_(
                              dim=1,
                              index=mapping,
                              src=torch.ones_like(mapping, dtype=logical_loads.dtype),
                          )
                          loads_per_replica = torch.gather(
                              logical_loads / replica_counts.clamp_min(1),
                              dim=1,
                              index=mapping,
                          )
                          loads_per_rank = loads_per_replica.reshape(
                              logical_loads.shape[0], ep_size, -1
                          ).sum(dim=(0, 2))
                          mean_load = loads_per_rank.mean()
                          if mean_load == 0:
                              return 1.0
                          return (loads_per_rank.max() / mean_load).item()

                      current_imbalance = rank_load_imbalance(
                          eplb_model_state.physical_to_logical_map
                      )
                      proposed_imbalance = rank_load_imbalance(
                          new_physical_to_logical_map
                      )
                      relative_improvement = (
                          current_imbalance - proposed_imbalance
                      ) / current_imbalance
                      skip_rearrange = relative_improvement < 0.05
  ```

Comment (tlrmchlsmth) [reply vllm/distributed/eplb/eplb_state.py]: Trying to simplify + clean up this block of code before landing - @okorzh-amd and @AndreasKaratzas could you take a look?

Comment (tlrmchlsmth) [reply vllm/distributed/eplb/eplb_state.py]: We should conservatively skip rearranging on non-AMD hardware as well IMO, since it adds an allreduce.

**Review by mgoin — APPROVED** [review]
